cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java

```diff
@@ -107,6 +118,61 @@ public OperationController startOperation(OperationRunDetail detail) {
     return runtime.run(detail);
   }
 
+  /**
+   * Create a new pull operation. Inserts the run in DB and then send TMS message.
+   */
+  public OperationRun createPushOperation(String namespace, String runId, PushAppsRequest request, String principal)
+      throws IOException, TooManyRequestsException {
+    OperationMeta meta = OperationMeta.builder().setCreateTime(Instant.now()).build();
+    OperationRunId operationRunId = new OperationRunId(namespace, runId);
+    OperationRun run = OperationRun.builder()
+        .setRunId(runId)
+        .setMetadata(meta)
+        .setStatus(OperationRunStatus.STARTING)
+        .setType(OperationType.PUSH_APPS)
+        .build();
+    OperationRunDetail detail = OperationRunDetail.builder()
+        .setRun(run)
+        .setPrincipal(principal)
+        .setPushAppsRequest(request)
+        .setRunId(operationRunId)
+        .build();
+    TransactionRunners.run(transactionRunner, context -> {
+      validateOnlyOneGitOperationRunning(namespace, context);
+      getOperationRunStore(context).createOperationRun(operationRunId, detail);
+      statePublisher.publishStarting(operationRunId);
```

Comment:
note that it is possible to publish multiple starting messages if this transaction fails and is retried. Though I think the subscriber handles that case so it's not an issue.

Reply:
Yeah publishing multiple messages should not be an issue here as subscriber will effectively deduplicate it